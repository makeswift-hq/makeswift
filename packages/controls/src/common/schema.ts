import { z } from 'zod'

import { ResourceSchema } from '../resources'

const primitives = z.union([
  z.undefined(),
  z.null(),
  z.boolean(),
  z.number(),
  z.string(),
])

// See https://github.com/colinhacks/zod?tab=readme-ov-file#recursive-types
type Data = z.infer<typeof primitives> | Data[] | { [key: string]: Data }

export const data: z.ZodType<Data> = z.any()

export const colorData = z.object({
  swatchId: z.string(),
  alpha: z.number(),
})

export const resolvedColorData = z.object({
  swatch: ResourceSchema.swatch,
  alpha: z.number(),
})

export const deviceId = z.string()

export const deviceOverride = <T extends z.ZodTypeAny>(value: T) =>
  z.object({
    deviceId,
    value,
  })

export const responsiveValue = <T extends z.ZodTypeAny>(value: T) =>
  z.array(deviceOverride(value))

export const elementData = z.object({
  type: z.string(),
  key: z.string(),
  props: z.record(data),
})

export const elementReference = z.object({
  type: z.literal('reference'),
  key: z.string(),
  value: z.string(),
})

export const element = z.union([elementData, elementReference])
