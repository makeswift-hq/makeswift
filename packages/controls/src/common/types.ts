import { z } from 'zod'

import * as Schema from './schema'

export { type BoxModel } from 'css-box-model'

export type Data = z.infer<typeof Schema.data>
export type ColorData = z.infer<typeof Schema.colorData>
export type ResolvedColorData = z.infer<typeof Schema.resolvedColorData>
export type Device = z.infer<typeof Schema.deviceId>
export type DeviceOverride<T> = { deviceId: Device; value: T }
export type ResponsiveValue<T> = DeviceOverride<T>[]
export type ResponsiveValueType<T> =
  T extends ResponsiveValue<infer U> ? U : never

export type ElementData = z.infer<typeof Schema.elementData>
export type ElementReference = z.infer<typeof Schema.elementReference>
export type Element = z.infer<typeof Schema.element>

export function isElementReference(
  element: Element,
): element is ElementReference {
  return !('props' in element)
}

export const ControlDataTypeKey = '@@makeswift/type'
