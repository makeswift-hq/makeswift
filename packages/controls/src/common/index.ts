export * from './data-type-key'
export * from './types'