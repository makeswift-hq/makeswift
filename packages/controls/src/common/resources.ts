export type Swatch = {
  id: string
  hue: number
  saturation: number
  lightness: number
}
