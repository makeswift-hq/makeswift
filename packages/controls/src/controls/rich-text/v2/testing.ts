import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
