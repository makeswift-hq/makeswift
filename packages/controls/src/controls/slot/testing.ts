import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
