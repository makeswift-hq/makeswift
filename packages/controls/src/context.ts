import { type Data, type Element, type ElementData } from './common/types'

export type TranslationDto = Record<string, Data>
export type MergeTranslatableDataContext = {
  translatedData: TranslationDto
  mergeTranslatedData: (node: Element) => Element
}

export type ReplacementContext = {
  elementHtmlIds: Set<string>
  elementKeys: Map<string, string>
  swatchIds: Map<string, string>
  fileIds: Map<string, string>
  typographyIds: Map<string, string>
  tableIds: Map<string, string>
  tableColumnIds: Map<string, string>
  pageIds: Map<string, string>
  globalElementIds: Map<string, string>
  globalElementData: Map<string, ElementData>
}

export type SerializableReplacementContext = {
  elementHtmlIds?: string[]
  elementKeys?: Record<string, string>
  swatchIds?: Record<string, string>
  fileIds?: Record<string, string>
  typographyIds?: Record<string, string>
  tableIds?: Record<string, string>
  tableColumnIds?: Record<string, string>
  pageIds?: Record<string, string>
  globalElementIds?: Record<string, string>
  globalElementData?: Record<string, ElementData>
}

export function createReplacementContext(
  context: SerializableReplacementContext,
): ReplacementContext {
  const toMap = <V>(record?: Record<string, V>) =>
    new Map(Object.entries(record ?? {}))

  return {
    elementHtmlIds: new Set(context.elementHtmlIds),
    elementKeys: toMap(context.elementKeys),
    swatchIds: toMap(context.swatchIds),
    fileIds: toMap(context.fileIds),
    typographyIds: toMap(context.typographyIds),
    tableIds: toMap(context.tableIds),
    tableColumnIds: toMap(context.tableColumnIds),
    pageIds: toMap(context.pageIds),
    globalElementIds: toMap(context.globalElementIds),
    globalElementData: toMap(context.globalElementData),
  }
}

export type CopyContext = {
  replacementContext: ReplacementContext
  copyElement: (node: Element) => Element
}

export type MergeContext = {
  mergeElement(a: Element, b: Element): Element
}
