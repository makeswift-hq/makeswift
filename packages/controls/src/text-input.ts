import { ControlDataTypeKey } from './common'

export const TextInputControlDataTypeKey = ControlDataTypeKey

export const TextInputControlDataTypeValueV1 = 'text-input::v1'

export type TextInputControlDataV0 = string

export type TextInputControlDataV1 = {
  [ControlDataTypeKey]: typeof TextInputControlDataTypeValueV1
  value: string
}

export type TextInputControlData = TextInputControlDataV0 | TextInputControlDataV1

export const TextInputControlType = 'makeswift::controls::text-input'

type TextInputControlConfig = {
  label?: string
  defaultValue?: string
}

export type TextInputControlDefinition<C extends TextInputControlConfig = TextInputControlConfig> =
  {
    type: typeof TextInputControlType
    config: C
    version?: 1
  }

export function TextInput<C extends TextInputControlConfig>(
  config: C = {} as C,
): TextInputControlDefinition<C> {
  return { type: TextInputControlType, config, version: 1 }
}
