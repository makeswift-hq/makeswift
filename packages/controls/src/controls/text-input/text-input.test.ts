import { testDefinition } from '../../testing/test-definition'

import { ControlDataTypeKey } from '../../common'
import { MergeTranslatableDataContext, TranslationDto } from '../../context'

import { TextInput, TextInputDefinition } from './text-input'

describe('TextInput', () => {
  describe('constructor', () => {
    test.each(['bert', 'ernie', undefined])(
      'call with default value `%s` returns versioned definition',
      (value) => {
        expect(
          TextInput({
            label: 'TextInput',
            defaultValue: value,
          }),
        ).toMatchSnapshot()
      },
    )

    test('disallows extraneous properties', () => {
      TextInput({
        label: undefined,
        defaultValue: undefined,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: TextInputDefinition) {}
    assignTest(TextInput())
    assignTest(TextInput({ label: 'TextInput' }))
    assignTest(TextInput({ defaultValue: 'text' }))
    assignTest(TextInput({ defaultValue: 'text' }))
    assignTest(TextInput({ defaultValue: 'text' as string }))
    assignTest(TextInput({ label: 'TextInput', defaultValue: undefined }))
    assignTest(TextInput({ label: undefined, defaultValue: undefined }))
  })

  describe('getTranslatableData', () => {
    test.each([
      'cookie monster',
      {
        [ControlDataTypeKey]: 'text-input::v1' as const,
        value: 'elmo',
      },
      undefined,
    ])('returns `%s` when data is `%s`', (data) => {
      const def = TextInput({ label: 'TextInput' })
      expect(def.getTranslatableData(data)).toBe(data)
    })
  })

  describe('mergeTranslatedData', () => {
    const translationData: TranslationDto = {
      key: 'monstre de biscuits',
    }

    const mergeContext: MergeTranslatableDataContext = {
      translatedData: translationData,
      mergeTranslatedData: (el) => el,
    }

    const def = TextInput({ label: 'TextInput' })

    test('returns translated data when translated data is non-nullish', () => {
      const dataV0 = 'cookie monster'
      expect(
        def.mergeTranslatedData(dataV0, translationData, mergeContext),
      ).toBe(translationData)

      const dataV1 = {
        [ControlDataTypeKey]: 'text-input::v1' as const,
        value: 'elmo',
      }
      expect(
        def.mergeTranslatedData(dataV1, translationData, mergeContext),
      ).toBe(translationData)
    })

    test('returns data when translated data is nullish', () => {
      const dataV0 = 'cookie monster'

      const dataV1 = {
        [ControlDataTypeKey]: 'text-input::v1' as const,
        value: 'elmo',
      }

      expect(def.mergeTranslatedData(dataV0, undefined, mergeContext)).toBe(
        dataV0,
      )

      expect(def.mergeTranslatedData(dataV1, undefined, mergeContext)).toBe(
        dataV1,
      )
    })
  })

  const invalidValues = [null, false, 5, []]

  testDefinition(
    TextInput({ defaultValue: 'elmo', label: 'visible' }),
    ['bert', 'ernie'],
    invalidValues,
  )

  testDefinition(
    TextInput({ label: 'TextInput' }),
    ['big bird', 'cookie monster', undefined],
    invalidValues,
  )
})
