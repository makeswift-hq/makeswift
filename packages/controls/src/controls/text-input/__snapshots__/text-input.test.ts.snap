// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput constructor call with default value \`bert\` returns versioned definition 1`] = `
TextInputDefinition {
  "config": {
    "defaultValue": "bert",
    "label": "TextInput",
  },
  "version": 1,
}
`;

exports[`TextInput constructor call with default value \`ernie\` returns versioned definition 1`] = `
TextInputDefinition {
  "config": {
    "defaultValue": "ernie",
    "label": "TextInput",
  },
  "version": 1,
}
`;

exports[`TextInput constructor call with default value \`undefined\` returns versioned definition 1`] = `
TextInputDefinition {
  "config": {
    "defaultValue": undefined,
    "label": "TextInput",
  },
  "version": 1,
}
`;

exports[`TextInput definition w/ config {"defaultValue":"elmo","label":"visible"} toData returns v1 data for \`bert\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-input::v1",
  "value": "bert",
}
`;

exports[`TextInput definition w/ config {"defaultValue":"elmo","label":"visible"} toData returns v1 data for \`ernie\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-input::v1",
  "value": "ernie",
}
`;

exports[`TextInput definition w/ config {"label":"TextInput"} toData returns v1 data for \`big bird\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-input::v1",
  "value": "big bird",
}
`;

exports[`TextInput definition w/ config {"label":"TextInput"} toData returns v1 data for \`cookie monster\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-input::v1",
  "value": "cookie monster",
}
`;

exports[`TextInput definition w/ config {"label":"TextInput"} toData returns v1 data for \`undefined\` when definition version is 1 1`] = `undefined`;
